_episodes/03-communicating-data.md

```diff
@@ -23,91 +23,257 @@ parallelisation often relies on how you communicate data.
+none of this communication happens for free, either. With every message sent, there is an overhead which impacts the
+performance of your program. Often we won't notice this overhead, as it is usually quite small. But if we communicate
+large amounts data or small amounts too often, those small overheads rapidly add up into a noticeable performance hit.
 
 To get an idea of how communication typically happens, imagine we have two ranks: rank A and rank B. If rank A wants to
 send data to rank B (e.g., point-to-point), it must first call the appropriate MPI send function which puts that data
-into an internal *buffer*; sometimes known as the send buffer or envelope. Once the data is in the buffer, MPI figures
-out how to route the message to rank B (usually over a network) and sends it to B. To receive the data, rank B must call
-a data receiving function which will listen for any messages being sent. When the message has been successfully routed
-and the data transfer complete, rank B sends an acknowledgement back to rank A to say that the transfer has finished,
-similarly to how read receipts work in e-mails and instant messages.
+into an internal *buffer*; known as the **send buffer** or **envelope**. Once the data is in the buffer, MPI figures
+out how to route the message to rank B (usually over a network) and then sends it to B. To receive the data, rank B must
+call a data receiving function which will listen for messages being sent to it. In some cases, rank B will then send
+an acknowledgement to say that the transfer has finished, similar to read receipts in e-mails and instant messages.
 
 > ## Check your understanding
 >
-> In an imaginary simulation, each rank is responsible for calculating the physical properties for a subset of
-> cells on a larger simulation grid. Another calculation, however, needs to know the average of, for example, the
-> temperature for the subset of cells for each rank. What approaches could you use to share this data?
+> Consider a simulation where each rank calculates the physical properties for a subset of cells on a very large
+> grid of points. However, the next step of the calculation needs to know the average temperature across the entire grid
+> of points. How would you use calculate the average temperature?
 >
 > > ## Solution
 > >
 > > There are multiple ways to approach this situation, but the most efficient approach would be to use collective
-> > operations to send the average temperature to a root rank (or all ranks) to perform the final calculation. You can,
-> > of course, also use a point-to-point pattern, but it would be less efficient.
+> > operations to send the average temperature to a main rank which performs the final calculation. You can, of course,
+> > also use a point-to-point pattern, but it would be less efficient, especially with a large number of ranks.
+> >
+> > If you weren't do this in parallel, or instead using shared-memory parallelism, you wouldn't need to do any

```

Comment:
Perhaps add OpenMP as an example of shared memory parallelism, e.g. "...(e.g. using OpenMP)..."?